File: 22/py/22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def backTracking(cur_str :str, left_num :int, right_num :int) :
            if len(cur_str) == 2*n :
                result.append(cur_str)
                return

            if left_num < n :
                backTracking(cur_str + "(", left_num + 1, right_num)

            if right_num < left_num :
                backTracking(cur_str + ")", left_num, right_num + 1)

        backTracking("", 0, 0)

        return result

File: 22/go/22.go
// Backtracking

// number of left "(" >= number of right ")" always

func generateParenthesis(n int) []string {
    var result []string

    var Backtracking func(curStr string, leftNum int, rightNum int)

    Backtracking = func(curStr string, leftNum int, rightNum int) {
        if len(curStr) == 2 * n {
            result = append(result, curStr)
            return
        }

        if leftNum < n {
            Backtracking(curStr + "(", leftNum + 1, rightNum)
        }

        if rightNum < leftNum {
            Backtracking(curStr + ")", leftNum, rightNum + 1)
        }
    }

    Backtracking("", 0, 0)

    return result
}